File: analytics/src/test/java/io/appmetrica/analytics/impl/billing/BillingMonitorWrapperTest.java
package io.appmetrica.analytics.impl.billing;

import android.content.Context;
import io.appmetrica.analytics.billinginterface.internal.BillingType;
import io.appmetrica.analytics.billinginterface.internal.config.BillingConfig;
import io.appmetrica.analytics.billinginterface.internal.monitor.BillingMonitor;
import io.appmetrica.analytics.billinginterface.internal.storage.BillingInfoSender;
import io.appmetrica.analytics.billinginterface.internal.storage.BillingInfoStorage;
import io.appmetrica.analytics.coreapi.internal.servicecomponents.applicationstate.ApplicationState;
import io.appmetrica.analytics.coreapi.internal.servicecomponents.applicationstate.ApplicationStateObserver;
import io.appmetrica.analytics.coreapi.internal.servicecomponents.applicationstate.ApplicationStateProvider;
import io.appmetrica.analytics.impl.startup.StartupState;
import io.appmetrica.analytics.testutils.CommonTest;
import io.appmetrica.analytics.testutils.TestUtils;
import java.util.concurrent.Executor;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyZeroInteractions;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
public class BillingMonitorWrapperTest extends CommonTest {

    @Mock
    private Context context;
    @Mock
    private Executor workingExecutor;
    @Mock
    private Executor uiExecutor;
    private final BillingType billingType = BillingType.LIBRARY_V3;
    @Mock
    private BillingInfoStorage storage;
    @Mock
    private BillingInfoSender sender;
    @Mock
    private ApplicationStateProvider applicationStateProvider;
    @Mock
    private BillingMonitorProvider billingMonitorProvider;
    @Mock
    private BillingMonitor billingMonitor;
    @Mock
    private BillingConfig config;
    @Captor
    private ArgumentCaptor<ApplicationStateObserver> observerCaptor;
    private StartupState startupState;
    private BillingMonitorWrapper wrapper;

    @Before
    public void setUp() {
        MockitoAnnotations.openMocks(this);
        startupState = TestUtils.createDefaultStartupStateBuilder()
                .withAutoInappCollectingConfig(config)
                .build();
        when(billingMonitorProvider.get(context, workingExecutor, uiExecutor, billingType, storage, sender))
                .thenReturn(billingMonitor);
        when(applicationStateProvider.registerStickyObserver(any(ApplicationStateObserver.class)))
                .thenReturn(ApplicationState.UNKNOWN);
        wrapper = new BillingMonitorWrapper(
                context,
                workingExecutor,
                uiExecutor,
                billingType,
                storage,
                sender,
                applicationStateProvider,
                billingMonitorProvider
        );
    }

    @Test
    public void startWatchingFeatureDisabled() {
        wrapper.maybeStartWatching(startupState, false);
        verifyZeroInteractions(billingMonitorProvider);
        verifyZeroInteractions(applicationStateProvider);
    }

    @Test
    public void startWatchingFeatureNull() {
        wrapper.maybeStartWatching(startupState, null);
        verify(billingMonitorProvider).get(context, workingExecutor, uiExecutor, billingType, storage, sender);
        verify(billingMonitor).onBillingConfigChanged(config);
        verify(applicationStateProvider).registerStickyObserver(any(ApplicationStateObserver.class));
    }

    @Test
    public void startWatchingFeatureEnabled() {
        wrapper.maybeStartWatching(startupState, true);
        verify(billingMonitorProvider).get(context, workingExecutor, uiExecutor, billingType, storage, sender);
        verify(billingMonitor).onBillingConfigChanged(config);
        verify(applicationStateProvider).registerStickyObserver(any(ApplicationStateObserver.class));
    }

    @Test
    public void onStartupChangedHasMonitor() {
        wrapper.maybeStartWatching(startupState, null);
        BillingConfig newConfig = mock(BillingConfig.class);
        wrapper.onStartupStateChanged(TestUtils.createDefaultStartupStateBuilder()
                .withAutoInappCollectingConfig(newConfig)
                .build());
        verify(billingMonitor).onBillingConfigChanged(newConfig);
    }

    @Test
    public void onStartupChangedNoMonitor() {
        wrapper.maybeStartWatching(startupState, false);
        BillingConfig newConfig = mock(BillingConfig.class);
        wrapper.onStartupStateChanged(TestUtils.createDefaultStartupStateBuilder()
                .withAutoInappCollectingConfig(newConfig)
                .build());
    }

    @Test
    public void stateIsImmediatelyUnknownHasMonitor() throws Throwable {
        when(applicationStateProvider.registerStickyObserver(any(ApplicationStateObserver.class)))
                .thenReturn(ApplicationState.UNKNOWN);
        wrapper.maybeStartWatching(startupState, true);
        verify(billingMonitor, never()).onSessionResumed();
    }

    @Test
    public void stateIsImmediatelyVisibleHasMonitor() throws Throwable {
        when(applicationStateProvider.registerStickyObserver(any(ApplicationStateObserver.class)))
                .thenReturn(ApplicationState.VISIBLE);
        wrapper.maybeStartWatching(startupState, true);
        verify(billingMonitor).onSessionResumed();
    }

    @Test
    public void stateIsImmediatelyBackgroundHasMonitor() throws Throwable {
        when(applicationStateProvider.registerStickyObserver(any(ApplicationStateObserver.class)))
                .thenReturn(ApplicationState.BACKGROUND);
        wrapper.maybeStartWatching(startupState, true);
        verify(billingMonitor, never()).onSessionResumed();
    }

    @Test
    public void stateIsImmediatelyUnknownNoMonitor() {
        when(applicationStateProvider.registerStickyObserver(any(ApplicationStateObserver.class)))
                .thenReturn(ApplicationState.UNKNOWN);
        wrapper.maybeStartWatching(startupState, false);
    }

    @Test
    public void stateIsImmediatelyVisibleNoMonitor() {
        when(applicationStateProvider.registerStickyObserver(any(ApplicationStateObserver.class)))
                .thenReturn(ApplicationState.VISIBLE);
        wrapper.maybeStartWatching(startupState, false);
    }

    @Test
    public void stateIsImmediatelyBackgroundNoMonitor() {
        when(applicationStateProvider.registerStickyObserver(any(ApplicationStateObserver.class)))
                .thenReturn(ApplicationState.BACKGROUND);
        wrapper.maybeStartWatching(startupState, false);
    }

    @Test
    public void stateChangedToUnknownHasMonitor() throws Throwable {
        wrapper.maybeStartWatching(startupState, true);
        verify(applicationStateProvider).registerStickyObserver(observerCaptor.capture());
        observerCaptor.getValue().onApplicationStateChanged(ApplicationState.UNKNOWN);
        verify(billingMonitor, never()).onSessionResumed();
    }

    @Test
    public void stateChangedToVisibleHasMonitor() throws Throwable {
        wrapper.maybeStartWatching(startupState, true);
        verify(applicationStateProvider).registerStickyObserver(observerCaptor.capture());
        observerCaptor.getValue().onApplicationStateChanged(ApplicationState.VISIBLE);
        verify(billingMonitor).onSessionResumed();
    }

    @Test
    public void stateChangedToBackgroundHasMonitor() throws Throwable {
        wrapper.maybeStartWatching(startupState, true);
        verify(applicationStateProvider).registerStickyObserver(observerCaptor.capture());
        observerCaptor.getValue().onApplicationStateChanged(ApplicationState.BACKGROUND);
        verify(billingMonitor, never()).onSessionResumed();
    }

}

File: billing/src/test/java/io/appmetrica/analytics/billing/impl/BillingMonitorWrapperTest.kt
package io.appmetrica.analytics.billing.impl

import android.content.Context
import io.appmetrica.analytics.billing.impl.config.service.model.ServiceSideBillingConfig
import io.appmetrica.analytics.billing.impl.config.service.model.ServiceSideRemoteBillingConfig
import io.appmetrica.analytics.billing.impl.protobuf.client.AutoInappCollectingInfoProto
import io.appmetrica.analytics.billing.impl.sender.BillingInfoSenderImpl
import io.appmetrica.analytics.billing.impl.storage.AutoInappCollectingInfo
import io.appmetrica.analytics.billing.impl.storage.BillingInfoStorageImpl
import io.appmetrica.analytics.billinginterface.internal.BillingType
import io.appmetrica.analytics.billinginterface.internal.config.BillingConfig
import io.appmetrica.analytics.billinginterface.internal.monitor.BillingMonitor
import io.appmetrica.analytics.coreapi.internal.data.ProtobufConverter
import io.appmetrica.analytics.coreapi.internal.data.ProtobufStateSerializer
import io.appmetrica.analytics.coreapi.internal.executors.IHandlerExecutor
import io.appmetrica.analytics.coreapi.internal.servicecomponents.ServiceComponentModuleConfig
import io.appmetrica.analytics.coreapi.internal.servicecomponents.ServiceComponentModuleReporter
import io.appmetrica.analytics.coreapi.internal.servicecomponents.ServiceModuleReporterComponentContext
import io.appmetrica.analytics.coreapi.internal.servicecomponents.applicationstate.ApplicationState
import io.appmetrica.analytics.coreapi.internal.servicecomponents.applicationstate.ApplicationStateProvider
import io.appmetrica.analytics.modulesapi.internal.common.ExecutorProvider
import io.appmetrica.analytics.modulesapi.internal.service.ServiceContext
import io.appmetrica.analytics.modulesapi.internal.service.ServiceStorageProvider
import io.appmetrica.analytics.testutils.CommonTest
import io.appmetrica.analytics.testutils.constructionRule
import io.appmetrica.analytics.testutils.on
import io.appmetrica.analytics.testutils.staticRule
import org.assertj.core.api.Assertions.assertThat
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.kotlin.any
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.eq
import org.mockito.kotlin.mock
import org.mockito.kotlin.never
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
internal class BillingMonitorWrapperTest : CommonTest() {

    private val context: Context = mock()
    private val defaultExecutor: IHandlerExecutor = mock()
    private val uiExecutor: IHandlerExecutor = mock()
    private val reportRunnableExecutor: IHandlerExecutor = mock()
    private val executorProvider: ExecutorProvider = mock {
        on { getDefaultExecutor() } doReturn defaultExecutor
        on { getUiExecutor() } doReturn uiExecutor
        on { getReportRunnableExecutor() } doReturn reportRunnableExecutor
    }
    private val applicationStateProvider: ApplicationStateProvider = mock()
    private val serviceStorageProvider: ServiceStorageProvider = mock {
        on {
            createBinaryStateStorageFactory(
                any(),
                any<ProtobufStateSerializer<AutoInappCollectingInfoProto.AutoInappCollectingInfo>>(),
                any<ProtobufConverter<AutoInappCollectingInfo, AutoInappCollectingInfoProto.AutoInappCollectingInfo>>()
            )
        } doReturn mock()
    }
    private val serviceContext: ServiceContext = mock {
        on { context } doReturn context
        on { executorProvider } doReturn executorProvider
        on { applicationStateProvider } doReturn applicationStateProvider
        on { serviceStorageProvider } doReturn serviceStorageProvider
    }

    private val serviceComponentModuleReporter: ServiceComponentModuleReporter = mock()
    private val serviceComponentModuleConfig: ServiceComponentModuleConfig = mock {
        on { isRevenueAutoTrackingEnabled() } doReturn true
    }
    private val serviceModuleReporterComponentContext: ServiceModuleReporterComponentContext = mock {
        on { reporter } doReturn serviceComponentModuleReporter
        on { config } doReturn serviceComponentModuleConfig
    }

    private val serviceSideRemoteBillingConfig = ServiceSideRemoteBillingConfig(
        enabled = true,
        config = ServiceSideBillingConfig(
            sendFrequencySeconds = 42,
            firstCollectingInappMaxAgeSeconds = 4242
        )
    )

    @get:Rule
    val billingTypeDetectorRule = staticRule<BillingTypeDetector> {
        on { BillingTypeDetector.getBillingType() } doReturn BillingType.LIBRARY_V8
    }

    @get:Rule
    val billingInfoStorageImplRule = constructionRule<BillingInfoStorageImpl>()

    @get:Rule
    val billingInfoSenderImplRule = constructionRule<BillingInfoSenderImpl>()

    private val billingMonitor: BillingMonitor = mock()

    @get:Rule
    val billingMonitorProviderRule = constructionRule<BillingMonitorProvider> {
        on {
            get(
                eq(context),
                eq(defaultExecutor),
                eq(uiExecutor),
                eq(BillingType.LIBRARY_V8),
                any(),
                any()
            )
        } doReturn billingMonitor
    }

    private val wrapper = BillingMonitorWrapper(
        serviceContext,
        serviceSideRemoteBillingConfig
    )

    @Test
    fun onMainReporterCreated() {
        whenever(applicationStateProvider.registerStickyObserver(any())).thenReturn(ApplicationState.VISIBLE)

        wrapper.onMainReporterCreated(serviceModuleReporterComponentContext)

        assertThat(billingMonitorProviderRule.constructionMock.constructed()).isNotEmpty()
        verify(billingMonitor).onBillingConfigChanged(BillingConfig(42, 4242))
        verify(applicationStateProvider).registerStickyObserver(any())
        verify(billingMonitor).onSessionResumed()
    }

    @Test
    fun onMainReporterCreatedIfNotVisible() {
        whenever(applicationStateProvider.registerStickyObserver(any())).thenReturn(ApplicationState.BACKGROUND)

        wrapper.onMainReporterCreated(serviceModuleReporterComponentContext)

        assertThat(billingMonitorProviderRule.constructionMock.constructed()).isNotEmpty()
        verify(billingMonitor).onBillingConfigChanged(BillingConfig(42, 4242))
        verify(applicationStateProvider).registerStickyObserver(any())
        verify(billingMonitor, never()).onSessionResumed()
    }

    @Test
    fun onMainReporterCreatedIfNoConfig() {
        whenever(applicationStateProvider.registerStickyObserver(any())).thenReturn(ApplicationState.VISIBLE)
        val wrapper = BillingMonitorWrapper(serviceContext, null)

        wrapper.onMainReporterCreated(serviceModuleReporterComponentContext)

        assertThat(billingMonitorProviderRule.constructionMock.constructed()).isNotEmpty()
        verify(billingMonitor, never()).onBillingConfigChanged(any())
        verify(applicationStateProvider).registerStickyObserver(any())
        verify(billingMonitor).onSessionResumed()
    }

    @Test
    fun updateConfig() {
        whenever(applicationStateProvider.registerStickyObserver(any())).thenReturn(ApplicationState.VISIBLE)

        wrapper.onMainReporterCreated(serviceModuleReporterComponentContext)

        assertThat(billingMonitorProviderRule.constructionMock.constructed()).isNotEmpty()
        verify(billingMonitor).onBillingConfigChanged(BillingConfig(42, 4242))
        verify(applicationStateProvider).registerStickyObserver(any())
        verify(billingMonitor).onSessionResumed()

        val newServiceSideRemoteBillingConfig = ServiceSideRemoteBillingConfig(
            enabled = true,
            config = ServiceSideBillingConfig(
                sendFrequencySeconds = 43,
                firstCollectingInappMaxAgeSeconds = 4343
            )
        )
        wrapper.updateConfig(newServiceSideRemoteBillingConfig)

        verify(billingMonitor).onBillingConfigChanged(BillingConfig(43, 4343))
    }

    @Test
    fun onMainReporterCreatedIfRevenueAutoTrackingIsDisabled() {
        whenever(serviceComponentModuleConfig.isRevenueAutoTrackingEnabled()).thenReturn(false)
        wrapper.onMainReporterCreated(serviceModuleReporterComponentContext)

        assertThat(billingMonitorProviderRule.constructionMock.constructed()).isEmpty()
    }
}
